python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,268 @@
+class WorkflowStorage:
+    """Access workflow in storage. This is a high-level function,
+    which does not care about the underlining storage implementation."""
+
+    def __init__(self,
+                 workflow_id: Optional[str] = None,
+                 store: Optional[storage.Storage] = None):
+        if workflow_id is None:
+            context = workflow_context.get_workflow_step_context()
+            workflow_id = context.workflow_id
+        if store is None:
+            store = storage.get_global_storage()
+        self._storage = store
+        self._workflow_dir = pathlib.Path(workflow_id)
+        self._steps_dir = self._workflow_dir / STEPS_DIR
+        self._objects_dir = self._workflow_dir / OBJECTS_DIR
+
+    def write_step_inputs(self, step_id: StepID, inputs: WorkflowInputs):
+        """Save workflow inputs."""
+        args = inputs.args
+        f = inputs.func_body
+        metadata = {
+            "step_id": inputs.step_id,
+            "name": f.__module__ + "." + f.__qualname__,
+            "object_refs": inputs.object_refs,
+            "workflows": inputs.workflows,
+        }
+        with serialization_context.workflow_args_keeping_context():
+            # TODO(suquark): in the future we should write to storage directly
+            # with plasma store object in memory.
+            args_obj = ray.get(args)
+        step_dir = self._steps_dir / step_id
+        if not self._storage.directory_exists(step_dir):
+            self._storage.create_directory(step_dir)
+        self._storage.write_json_atomic(metadata,
+                                        step_dir / STEP_INPUTS_METADATA)
+        self._storage.write_object_atomic(f, step_dir / STEP_FUNC_BODY)
+        self._storage.write_object_atomic(args_obj, step_dir / STEP_ARGS)
+
+    def write_step_output_metadata(self, step_id: StepID,
+                                   metadata: Dict[str, Any]) -> None:
+        """Save the output metadata json."""
+        step_dir = self._steps_dir / step_id
+        self._storage.write_json_atomic(metadata,
+                                        step_dir / STEP_OUTPUTS_METADATA)
+
+    def update_output_forward(self, forward_output_to: StepID,
+                              output_step_id: StepID) -> None:
+        """Update output forward. The output of 'output_step_id' should
+        forward to the step 'forward_output_to'. When resume from
+        'forward_output_to' step, that step can directly read
+        the output of 'output_step_id'.
+
+        Args:
+            forward_output_to: step 'forward_output_to'.
+            output_step_id: step 'output_step_id'.
+        """
+        if forward_output_to == "":
+            # actually it equals to 'self._workflow_dir / STEPS_DIR / ""',
+            # but we won't use the trick here because it is not obvious.
+            target_dir = self._workflow_dir / STEPS_DIR
+        else:
+            target_dir = self._workflow_dir / STEPS_DIR / forward_output_to
```

Comment:
In general can't this have multiple forwarding entries? Imagine a step has multiple inputs each of which can be dynamically expanded:

f.step(input1, input2)

Then, we'd need to separately track the output step id of input1 and input2.

Reply:
I think forwarding entries are used to shorten the lookup of nested workflows. Your case are about inputs, and we do not need to forward inputs. This is because:

1. if `f` finishes and checkpointed successfully, we already have `f`'s output checkpointing file, so we do not need to check the inputs during recovery.
2. If `f` does not finish, to reconstruct the output, we have to load each of `f`'s input anyway (input1, input2). Because we always forward the result of nested workflows, `input1` & `input2`'s output can be read using a single operation each.

Reply:
I see, so just to make sure I understand, in this case:

```
@workflow.step
def loop_expansion(i)
    if i == 0:
        return 0
    else:
        return loop_expansion.step(i - 1)

loop_result1 = loop_expansion.step(10000)
loop_result2 = loop_expansion.step(10000)
final_result = consume.step(loop_result1, loop_result2)

workflow.run(final_result)
```

The dag looks like this:

```
loop1 -\
        --- consume --- output
loop2 -/
```

On each loop step, we update the output_forward.json in loop1/loop2 to point to the end of the chain.

When we recover, we walk back from output => consume => loop1 / loop2, lookup the forwarding json, and jump to the end of the loop, is that right?

Reply:
Another question: assuming the above, then why do we sometimes need the forward.json in the root dir? Can't it always be in the step dir?

Reply:
Your understanding is correct. We need `forward.json` because currently we treat the root dir just as a normal workflow step: when you run `workflow.run(some_step.step())`, it is almost equivalent to returning a nested workflow `some_step.step()` from a workflow step. The benefits of doing this is that we do not need to create two sets of checkpointing / workflow execution logic, one for `workflow.run` and one for nested workflows. Because of this, we also have ` forward.json` in the root dir.